src/internal/restart/restart.go

```diff
@@ -0,0 +1,138 @@
+// Package restart shuts down Pachyderm when someone wants it to.
+package restart
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/admindb"
+	"github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dbutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"go.uber.org/zap"
+)
+
+var startupTime = time.Now()
+
+// checkFn is a function that checks if pachd needs to restart right now, returning whether to
+// restart and the reason for restarting.
+type checkFn func(ctx context.Context) (restart bool, reason string, err error)
+
+// restartFn is a function that restarts Pachyderm for the provided reason.
+type restartFn func(ctx context.Context, reason string)
+
+// Restarter is an object that manages restarting Pachyderm.
+type Restarter struct {
+	ch      listenerChan
+	check   checkFn
+	restart restartFn
```

Comment:
I see why `check` and `restart` are members which happen to be functions rather than methods — it enables you to mutate state (in the database and filesystem, as well as of the running process itself) in a real `Restarter` and not do so in a test one.

It _might_ be worth spending a little bit of time considering if that ought to be reified into an interface, with a database+filesystem implementation and a no-op one.

Either way, it would probably be good to document in the future for the next reader.